presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.reservoirsample;
+
+import java.util.Arrays;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.DoubleStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class UnweightedDoubleReservoirSample
+        implements Cloneable
+{
+    protected int maxSamples;
+    protected long count;
+    double[] elements;
+
+    public UnweightedDoubleReservoirSample(int maxSamples)
+    {
+        this.maxSamples = maxSamples;
+        checkArguments();
```

Comment:
I don't quite understand this comment. This constant is in some obscure inner class in the source code of general Presto. How do we know how much memory is running on other installations of Presto? At our installations in 5 years? 

The type of constant you're suggesting doesn't seem to me to belong in the code, but rather by some tool used by someone installing and maintaining Presto. They can use cgroups for that without modifying the source code.

Reply:
@atavory 

> They can use cgroups for that without modifying the source code.

Presto is a multi-tenant system, e.g. a single JVM is processing multiple queries at the same time. Cgroup configuration can limit the total amount of memory used by all the queries on a single host, but can't be used to limit amount of memory consumed by individual queries. Hence, it is important to ensure that no single query can use all of the memory on a host starving other queries and causing instability. This is the main reason why Presto cannot support UDFs that use external resources (make network calls, read from disk, allocate large chunks of memory).

It might be reasonable to make `MAX_SAMPLES_LIMIT` value configurable, however, it would be very hard for the system administrator to find the right value. Allowing any query to use way more memory than others increases the risk of having reliability problems.

There must be a limit. The default value for that limit should be such that it makes sense in most cases. If it is hard to come up with default value that works in 80% of use cases, then, a configuration option to change that limit might be helpful.

For the use cases you know about, what's the number of samples they require?